### ISSUE TITLE
XmlAttribute

### ISSUE THREAD

**Issue by wojtpl2** [wojtpl2/ExtendedXmlSerializer#6]
Add support for XmlAttribute

**Comment by Mike-EEE** [comment]
This would be great.  It would also be ideal to be able to mark a property as an "xml attribute" without the use of attributes.  That is, keep the POCO a POCO. :)  Additionally, it would be nice to have the ability to "auto-attribute" properties, so that strings are serialized as elements, but other scalar values (`DateTime`, `int`, `bool`, etc.) are serialized as attributes.

For example, I have [this class](https://github.com/Mike-EEE/Stash/blob/master/MsftBuild.Model/Contoso.Tasks/CustomTask.cs) that currently get serialized as such:

```
<CustomTask type="Contoso.Tasks.CustomTask">
	<SomeProperty>Hello World!</SomeProperty>
	<SomeDate>2016-11-23T00:00:00</SomeDate>
	<SomeNumber>123456789</SomeNumber>
</CustomTask>
```

Using a theoretical auto-attribute mode, it would be serialized as:

```
<CustomTask type="Contoso.Tasks.CustomTask" SomeDate="2016-11-23T00:00:00" SomeNumber="123456789">
	<SomeProperty>Hello World!</SomeProperty>
</CustomTask>
```

And of course, using a configuration object, it would be ideal to be able to further customize `SomeProperty` to make it an attribute as well.  Thank you for any consideration!

**Comment by FaustoNascimento** [comment]
All up for the XmlAttribute support, but not sure about the auto-attribute feature.

I think it should mirror whatever XmlSerializer does by default, but you should offer a way to configure which object types get represented as elements and which as attributes (with a default which mirrors that of XmlSerializer).

**Comment by Mike-EEE** [comment]
Indeed @FaustoNascimento we are providing an "alias" or "name" function for all properties/fields which can be done via the ConfigurationAPI and/or by metadata/attribute.  It will first look at what is configured via an `ExtendedXmlConfiguration` container, and if nothing is specified there, it will then look at the metadata (ala, classic `XmlSerializer`).

To be sure, the "auto-attribute" feature is a configuration command (or extension) that you have to explicitly call on the configuration container to enable.  So it's not fancy magic that is enabled by default, but rather it's fancy magic you have to explicitly invoke. :)

Definitely open to any additional feedback to this, especially since I am starting towards developing the final leg (ConfiguationAPI) today!

**Comment by wojtpl2** [comment]
I think that auto-attribute feature should be disabled default.

If you have class:

```C#
class X {
 prop A;

 [XmlAttribute("A")] //or even without changing name 
 prop B;
}
```
XmlSerializer will return:

```XML
<X B="">
 <A>sadf</A>
<X>
```

After migrate code to EXS and with auto-attribute feature enabled you will retrieve different xml. 
```XML
<X A="" B=""> // there is question about order attributes ;)
<X>
```

For me the most important question is:
Will we be able to read old xml (from XmlSerializer) if we set auto-attribute feature enabled by default. 
If not we can't do this.

**Comment by Mike-EEE** [comment]
Agreed, @wojtpl2.  The way you will get auto-attributes is something like this:

```
var configuration = new ExtendedXmlConfiguration().Extend(AutoAttributesExtension.Default); 
var serializer = confiuguration.Create();
```

We can also put this in an extension method, so something like:

```
var configuration = new ExtendedXmlConfiguration().EnableAutoAttributes(); 
var serializer = confiuguration.Create();
```

So it will not be default behavior but rather behavior that you have to call specifically.  👍

**Comment by wojtpl2** [comment]
Ok. That sounds great. 

We should have static or instance API for create serializer. Like this: https://github.com/AutoMapper/AutoMapper/wiki/Getting-started#how-do-i-use-automapper

What do you think ? It is easy to do and useful for simple case.

**Comment by Mike-EEE** [comment]
Ah, I think we had a discussion about this earlier.  Using a static method such as `Mapper.Initialize` is actually an anti-pattern.  In fact, I believe they spent an entire version in AutoMapper (from v3-4?) undoing that so that it resulted in the instance API that you refer to (I think, the "instance API" being much like my examples above?).

Consider why this is a bad pattern with xUnit.  Those tests are run in parallel.  If you have one static method where each test is assigning to a static method, then each method could possibly be using a different configuration depending on when it is executing. So:

1. Method1 assigns configuration.
2. Method2 assigns configuration
3. Method1 executes, but under Method2's configuration, and fails.

The other thing that I notice with AutoMapper is that they use delegates for configuration.  I have to say there are a lot of them in their API and I still to this day do not understand a lot of them.  It was more difficult to integrate and discover their API when upgrading to their latest version.

(FWIW/BTW, I use [AutoMapper for my Framework](https://github.com/DragonSpark/Framework) -- which, incidentally, has been put on hold since I started helping over here. 😄)

To me, this seems more, well, fluent:

```
var configuration = new ExtendedXmlConfiguration().UseAutoProperties()
			                          .UseFeatureOne()
			                          .UseFeatureTwo();
```

The other consideration with using delegates is that they can add to memory pressure with allocations:
https://blog.jetbrains.com/dotnet/2014/07/24/unusual-ways-of-boosting-up-app-performance-lambdas-and-linqs/

You will (hopefully) notice that I go to great lengths not to use delegates in EXS's current incarnation for exactly this reason.  It's part of the reason it is so fast. 😎

**Comment by Mike-EEE** [comment]
(I am definitely open to feedback here.  If you are really really wanting a static initializer method, you can most certainly add one!)

**Comment by wojtpl2** [comment]
I know that it is anti-pattern and it is not important for me. But I've seen that [Json.Net](http://www.newtonsoft.com/json) has static interface too:
`string json = JsonConvert.SerializeObject(product);`

In simple case static interface could be useful. We can go this way:
1. v2 without static interface.
2. If someone create issue for it we add static interface to v3.
What do you think?

**Comment by Mike-EEE** [comment]
Ah... well, sure.  If we want to make a shortcut/convenience method to quickly/easily serialize/deserialize object, then that is something worth looking into.  The problem is that it isn't best practice.  Or rather, it wouldn't be performant/fast because we would have to create a new `ExtendedXmlConfiguration` (and `ExtendedXmlSerializer`) each and every time.  So we would lose out on our `XmlConvert` caching with such an implementation.

But I do like your suggestion.  It will give me time to look up additional material that says statics are a very bad thing. :)

**Comment by Mike-EEE** [comment]
This should now be working as expected in the v2 branch.  If you encounter an issue with the current implementation and/or it does not behave as expected, please re-open this issue (or create a new one) and we will take it from there.

**Comment by Mike-EEE** [comment]
Pre-release deployment for this feature can be found at these feeds:
https://www.myget.org/feed/wojtpl2/package/nuget/ExtendedXmlSerializer
https://www.nuget.org/packages/ExtendedXmlSerializer/2.0.0-alpha
